File: src/engine/deduplication.ts
import { NormalizedEvent, AuditRecord } from '../types.js';

export interface LeadCluster {
  canonical_id: string;
  primary_email: string;
  primary_phone: string;
  primary_name: string;
  events: NormalizedEvent[];
  duplicate_count: number;
  merge_reasons: string[];
}

export function clusterAndDeduplicateEvents(
  events: NormalizedEvent[],
  existingAuditTrail?: AuditRecord[]
): {
  clusters: LeadCluster[];
  duplicateAuditRecords: AuditRecord[];
} {
  // We use a Disjoint-Set / Union-Find structure or deterministic clustering maps
  // to ensure that transitive linkages (e.g. event1 has email A & phone B, event2 has phone B)
  // or simple email/phone grouping are deterministic regardless of arrival order.

  const duplicateAuditRecords: AuditRecord[] = [];

  // Group events deterministically.
  // First, index by email, phone, and lead_id
  const emailToGroupIndex = new Map<string, number>();
  const phoneToGroupIndex = new Map<string, number>();
  const explicitIdToGroupIndex = new Map<string, number>();

  // To maintain absolute determinism regardless of arrival order:
  // Sort events first by a stable signature (e.g., event_id) before initial clustering evaluation
  const sortedEvents = [...events].sort((a, b) => a.event_id.localeCompare(b.event_id));

  // Disjoint set parent pointers
  const parent: number[] = [];
  function find(i: number): number {
    if (parent[i] === i) return i;
    parent[i] = find(parent[i]);
    return parent[i];
  }
  function union(i: number, j: number) {
    const rootI = find(i);
    const rootJ = find(j);
    if (rootI !== rootJ) {
      if (rootI < rootJ) {
        parent[rootJ] = rootI;
      } else {
        parent[rootI] = rootJ;
      }
    }
  }

  for (let i = 0; i < sortedEvents.length; i++) {
    parent[i] = i;
  }

  // Link events
  for (let i = 0; i < sortedEvents.length; i++) {
    const evt = sortedEvents[i];

    // Rule A: Normalized Email
    if (evt.email) {
      if (emailToGroupIndex.has(evt.email)) {
        union(i, emailToGroupIndex.get(evt.email)!);
      } else {
        emailToGroupIndex.set(evt.email, i);
      }
    }

    // Rule B: Normalized Phone (when phone is present)
    if (evt.phone) {
      if (phoneToGroupIndex.has(evt.phone)) {
        union(i, phoneToGroupIndex.get(evt.phone)!);
      } else {
        phoneToGroupIndex.set(evt.phone, i);
      }
    }

    // Rule C: If both email and phone are missing, match explicit lead_id if present
    if (!evt.email && !evt.phone && evt.lead_id) {
      if (explicitIdToGroupIndex.has(evt.lead_id)) {
        union(i, explicitIdToGroupIndex.get(evt.lead_id)!);
      } else {
        explicitIdToGroupIndex.set(evt.lead_id, i);
      }
    }
  }

  // Group events by root cluster
  const rootGroups = new Map<number, NormalizedEvent[]>();
  for (let i = 0; i < sortedEvents.length; i++) {
    const root = find(i);
    if (!rootGroups.has(root)) {
      rootGroups.set(root, []);
    }
    rootGroups.get(root)!.push(sortedEvents[i]);
  }

  const clusters: LeadCluster[] = [];

  for (const groupEvents of rootGroups.values()) {
    // Sort events deterministically within group by timestamp, then event_id
    groupEvents.sort((a, b) => {
      const timeDiff = a.timestamp.localeCompare(b.timestamp);
      if (timeDiff !== 0) return timeDiff;
      return a.event_id.localeCompare(b.event_id);
    });

    // Derive canonical lead ID deterministically
    // If any event already had a non-empty lead_id, select the lexicographically earliest one
    const explicitIds = groupEvents
      .map(e => e.lead_id)
      .filter(id => Boolean(id && id.trim()))
      .sort();

    // Primary email and phone
    const emails = groupEvents
      .map(e => e.email)
      .filter(Boolean)
      .sort();
    const phones = groupEvents
      .map(e => e.phone)
      .filter(Boolean)
      .sort();
    const names = groupEvents
      .map(e => e.name)
      .filter(Boolean)
      .sort((a, b) => b.length - a.length); // prefer fuller name

    const primaryEmail = emails[0] || '';
    const primaryPhone = phones[0] || '';
    const primaryName = names[0] || 'Anonymous Lead';

    let canonicalId = explicitIds[0];
    if (!canonicalId) {
      if (primaryEmail) {
        // Deterministic ID based on email prefix or sanitized email
        const sanitized = primaryEmail.replace(/[^a-zA-Z0-9]/g, '-').slice(0, 24);
        canonicalId = `lead-${sanitized}`;
      } else if (primaryPhone) {
        const sanitized = primaryPhone.replace(/[^0-9]/g, '');
        canonicalId = `lead-phone-${sanitized}`;
      } else {
        canonicalId = `lead-${groupEvents[0].event_id}`;
      }
    }

    const mergeReasons: string[] = [];
    const duplicateCount = Math.max(0, groupEvents.length - 1);

    // If there are duplicate events in the cluster, generate audit records
    if (groupEvents.length > 1) {
      const firstEvent = groupEvents[0];
      for (let k = 1; k < groupEvents.length; k++) {
        const dupEvent = groupEvents[k];
        let reason = '';
        if (dupEvent.email && dupEvent.email === firstEvent.email) {
          reason = `Merged with canonical lead ${canonicalId} (Normalized email matched ${dupEvent.email})`;
        } else if (dupEvent.phone && dupEvent.phone === firstEvent.phone) {
          reason = `Merged with canonical lead ${canonicalId} (Normalized phone matched ${dupEvent.phone})`;
        } else if (dupEvent.lead_id && dupEvent.lead_id === firstEvent.lead_id) {
          reason = `Merged with canonical lead ${canonicalId} (Explicit lead_id matched ${dupEvent.lead_id})`;
        } else {
          reason = `Merged with canonical lead ${canonicalId} based on linked identity parameters`;
        }

        mergeReasons.push(reason);

        duplicateAuditRecords.push({
          id: `audit-dup-${dupEvent.event_id}`,
          lead_id: canonicalId,
          event_id: dupEvent.event_id,
          decision_type: 'DUPLICATE_MERGE',
          previous_state: null,
          new_state: null,
          selected_campaign: dupEvent.campaign || null,
          candidate_campaigns: [dupEvent.campaign].filter(Boolean) as string[],
          reason,
          timestamp: dupEvent.timestamp,
        });
      }
    }

    clusters.push({
      canonical_id: canonicalId,
      primary_email: primaryEmail,
      primary_phone: primaryPhone,
      primary_name: primaryName,
      events: groupEvents,
      duplicate_count: duplicateCount,
      merge_reasons: mergeReasons,
    });
  }

  // Sort clusters deterministically by canonical_id
  clusters.sort((a, b) => a.canonical_id.localeCompare(b.canonical_id));

  return { clusters, duplicateAuditRecords };
}

File: src_py/deduplication.py
def cluster_and_deduplicate(events: list):
    sorted_events = sorted(events, key=lambda e: e['event_id'])
    n = len(sorted_events)
    parent = list(range(n))

    def find(i):
        if parent[i] == i:
            return i
        parent[i] = find(parent[i])
        return parent[i]

    def union(i, j):
        root_i = find(i)
        root_j = find(j)
        if root_i != root_j:
            if root_i < root_j:
                parent[root_j] = root_i
            else:
                parent[root_i] = root_j

    email_map = {}
    phone_map = {}
    lead_id_map = {}

    for i, evt in enumerate(sorted_events):
        if evt['email']:
            if evt['email'] in email_map:
                union(i, email_map[evt['email']])
            else:
                email_map[evt['email']] = i

        if evt['phone']:
            if evt['phone'] in phone_map:
                union(i, phone_map[evt['phone']])
            else:
                phone_map[evt['phone']] = i

        if not evt['email'] and not evt['phone'] and evt['lead_id']:
            if evt['lead_id'] in lead_id_map:
                union(i, lead_id_map[evt['lead_id']])
            else:
                lead_id_map[evt['lead_id']] = i

    groups = {}
    for i in range(n):
        root = find(i)
        if root not in groups:
            groups[root] = []
        groups[root].append(sorted_events[i])

    clusters = []
    duplicate_audit = []

    for group_events in groups.values():
        group_events.sort(key=lambda e: (e['timestamp'], e['event_id']))
        explicit_ids = sorted([e['lead_id'] for e in group_events if e.get('lead_id')])
        emails = sorted([e['email'] for e in group_events if e.get('email')])
        phones = sorted([e['phone'] for e in group_events if e.get('phone')])
        names = sorted([e['name'] for e in group_events if e.get('name')], key=lambda n: len(n), reverse=True)

        primary_email = emails[0] if emails else ""
        primary_phone = phones[0] if phones else ""
        primary_name = names[0] if names else "Anonymous Lead"

        canonical_id = explicit_ids[0] if explicit_ids else ""
        if not canonical_id:
            if primary_email:
                clean = "".join(c if c.isalnum() else "-" for c in primary_email)[:24]
                canonical_id = f"lead-{clean}"
            elif primary_phone:
                clean = "".join(c for c in primary_phone if c.isdigit())
                canonical_id = f"lead-phone-{clean}"
            else:
                canonical_id = f"lead-{group_events[0]['event_id']}"

        dup_count = max(0, len(group_events) - 1)
        if len(group_events) > 1:
            first_e = group_events[0]
            for dup_e in group_events[1:]:
                reason = f"Merged with canonical lead {canonical_id} (Linked match)"
                if dup_e.get('email') and dup_e.get('email') == first_e.get('email'):
                    reason = f"Merged with canonical lead {canonical_id} (Normalized email matched {dup_e['email']})"
                elif dup_e.get('phone') and dup_e.get('phone') == first_e.get('phone'):
                    reason = f"Merged with canonical lead {canonical_id} (Normalized phone matched {dup_e['phone']})"

                duplicate_audit.append({
                    "id": f"audit-dup-{dup_e['event_id']}",
                    "lead_id": canonical_id,
                    "event_id": dup_e['event_id'],
                    "decision_type": "DUPLICATE_MERGE",
                    "previous_state": None,
                    "new_state": None,
                    "selected_campaign": dup_e.get('campaign'),
                    "candidate_campaigns": [dup_e.get('campaign')] if dup_e.get('campaign') else [],
                    "reason": reason,
                    "timestamp": dup_e['timestamp']
                })

        clusters.append({
            "canonical_id": canonical_id,
            "primary_email": primary_email,
            "primary_phone": primary_phone,
            "primary_name": primary_name,
            "events": group_events,
            "duplicate_count": dup_count
        })

    clusters.sort(key=lambda c: c['canonical_id'])
    return clusters, duplicate_audit
